modules/api/php/endpoints/candidate/visit/electrophysiology/bidsfile/archive.class.inc

```diff
@@ -0,0 +1,150 @@
+ */
+namespace LORIS\api\Endpoints\Candidate\Visit\Electrophysiology\BidsFile;
+
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+use \LORIS\api\Endpoint;
+
+/**
+ * This class handles request to a recording's BIDS files' archive.
+ *
+ * @category API
+ * @package  LORIS
+ * @author   Cecile Madjar <cecile.madjar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://github.com/aces/Loris
+ */
+class Archive extends Endpoint implements \LORIS\Middleware\ETagCalculator
+{
+    /**
+     * A cache of the results of the endpoint, so that
+     * it doesn't need to be recalculated for the ETag and handler
+     */
+    private $_cache;
+
+    /**
+     * The requested recording
+     *
+     * @var \LORIS\Recording
+     */
+    private $_recording;
+
+    /**
+     * Contructor
+     *
+     * @param \LORIS\Recording $recording The requested recording
+     */
+    public function __construct(\LORIS\Recording $recording)
+    {
+        $this->_recording = $recording;
+    }
+
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return array('GET');
+    }
+
+    /**
+     * Versions of the LORIS API which are supported by this
+     * endpoint.
+     *
+     * @return array a list of supported API versions.
+     */
+    protected function supportedVersions() : array
+    {
+        return array(
+            'v0.0.3-dev',
+        );
+    }
+
+    /**
+     * Delegates the request to a private handler appropriate to the request's
+     * method.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        switch ($request->getMethod()) {
+        case 'GET':
+            return $this->_handleGET($request);
+
+        case 'OPTIONS':
+            return (new \LORIS\Http\Response())
+                ->withHeader('Allow', $this->allowedMethods());
+
+        default:
+            return new \LORIS\Http\Response\JSON\MethodNotAllowed(
+                $this->allowedMethods()
+            );
+        }
+    }
+
+    /**
+     * Create an array representation of this endpoint's response body
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleGET(ServerRequestInterface $request): ResponseInterface
+    {
+        if (isset($this->_cache)) {
+            return $this->_cache;
+        }
+
+        $info = $this->_recording->getArchiveFileInfo();
+
+        if (!$info->isFile()) {
+            error_log('file in database but not in file system');
+            return new \LORIS\Http\Response\JSON\NotFound();
```

Comment:
I'm on the fence about whether these should be NotFound or Internal Server Errors.. the fact that you feel the need to log it to the server error logs makes me lean towards internal server error.

(The error message should also include the filename if it sticks around..)

Reply:
so if I were to modify that, I would do the following?
1) remove the `error_log()` line
2) replace the `NotFound` by `InternalServerError` and append the message of the error?
```suggestion
            return new \LORIS\Http\Response\JSON\InternalServerError(
                'file ' . $info->getRealPath() . ' in database but not in file system'
            );
```

Reply:
If it's an internal server error, the error_log makes sense. If it's not found, the error_log doesn't. So either keep the error_log and change is from `NotFound` to `InternalServerError` (in which case the param passed to InternalServerError will be passed to the client), or remove the error_log and keep it NotFound. If the error_log stays, the error message in the server logs needs to have the file name so that the admin who reads the log knows what they need to investigate.

Reply:
ok. Thanks! All done.